awc/src/middleware/redirect.rs

```diff
@@ -0,0 +1,267 @@
+                let connector = self.connector.clone();
+                let max_redirect_times = self.max_redirect_times;
+
+                // backup the uri for reuse schema and authority.
+                let uri = match head {
+                    RequestHeadType::Owned(ref head) => head.uri.clone(),
+                    RequestHeadType::Rc(ref head, ..) => head.uri.clone(),
+                };
+
+                let fut = connector.call(ConnectRequest::Client(head, body, addr));
+
+                RedirectServiceFuture::Client {
+                    fut,
+                    max_redirect_times,
+                    uri: Some(uri),
+                    addr,
+                    connector: Some(connector),
+                }
+            }
+        }
+    }
+}
+
+pin_project_lite::pin_project! {
+    #[project = RedirectServiceProj]
+    pub enum RedirectServiceFuture<S>
+    where
+        S: Service<ConnectRequest, Response = ConnectResponse, Error = SendRequestError>,
+        S: 'static
+    {
+        Tunnel { #[pin] fut: S::Future },
+        Client {
+            #[pin]
+            fut: S::Future,
+            max_redirect_times: u8,
+            uri: Option<Uri>,
+            addr: Option<SocketAddr>,
+            connector: Option<Rc<S>>
+        }
+    }
+}
+
+impl<S> Future for RedirectServiceFuture<S>
+where
+    S: Service<ConnectRequest, Response = ConnectResponse, Error = SendRequestError> + 'static,
+{
+    type Output = Result<ConnectResponse, SendRequestError>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        match self.as_mut().project() {
+            RedirectServiceProj::Tunnel { fut } => fut.poll(cx),
+            RedirectServiceProj::Client {
+                fut,
+                max_redirect_times,
+                uri,
+                addr,
+                connector,
+            } => {
+                match ready!(fut.poll(cx))? {
+                    ConnectResponse::Client(res) => match res.head().status {
+                        StatusCode::MOVED_PERMANENTLY
+                        | StatusCode::FOUND
+                        | StatusCode::SEE_OTHER
+                        | StatusCode::TEMPORARY_REDIRECT
+                        | StatusCode::PERMANENT_REDIRECT
+                            if *max_redirect_times > 0 =>
+                        {
+                            let uri = uri.take().unwrap();
+
+                            // rebuild uri from the location header value.
+                            let uri = res
+                                .headers()
+                                .get(header::LOCATION)
+                                .map(|value| {
+                                    Uri::builder()
+                                        .scheme(uri.scheme().cloned().unwrap())
+                                        .authority(uri.authority().cloned().unwrap())
+                                        .path_and_query(value.as_bytes())
```

Comment:
Just wondering of why are we reusing scheme and authority if via location we would get full blown URL if so can be replaced with Url::parse:
```
        .map(Url::parse).transpose().map_err(SendRequestError::Url)?
```

Reply:
I failed to parse location to Uri type so this is what I end up with.

Reply:
Changed to parse the location to uri directly. And fallback to builder when the new uri lack schema and authority.